### DISCUSSION TITLE
Production deployment - horizontal scaling

### DISCUSSION THREAD

**Discussion by rossmcphee** [NRules/NRules#330] [Q&A]
We have implemented a NRules based calendar application that uses ISessions. Having built the ISession we use it to test how changes that users make on the UI affect the compliance of the resultant calendar to coverage rules.

Management have demanded that the application should scale horizontally under load.  However, since the ISessions are all stored in memory, the only alternatives I can see that would allow the application to support outward scaling are:

1. Stop using ISession and regenerate all the facts every time the user changes a fact.
2. Serialise the ISession and pass that between the worker instances (by Redis perhaps).  I see that you can serialise rules ([272](https://github.com/NRules/NRules/discussions/272)) but so far no support for ISession (I understand why not - it's mad to go to the effort of creating Rete trees in memory for speed and then serialize them but... management...).
3. Extract the NRules component into a stand alone service that the current App Service polls and insist that this service has to scale up not out.

Have I missed any other alternatives?

Has anyone got scaling out to work with NRule Session?

Are you considering adding Session serialisation or any other mechanism that would support scaling out?

**Comment by snikolayev** [comment]
@rossmcphee I don't think you necessarily need the session to be serializable to scale out your NRules-based application. But you can still scale out, though the solution would greatly depend on what it is you actually want to scale here. In other words, it depends on the aspect of your application that is growing.
There are several things at play with NRules that can scale:
1. The number of rules (you continue adding new rules all the time)
2. The number of facts in the session (the knowledge base of data used to make decisions is constantly growing)
3. The number of requests that your application is making to the session, to fire rules and get the inference results

From your description, it sounds like it's the 3rd case. The best way to approach this case, is to host the rules session in a separate service, and have a farm of these services, where different requests go to different instances. Each instance would have a copy of a session, and as facts change in the world, those changes are communicated to all those instances (e.g. via some broadcast messaging mechanism) and applied to all sessions. This way you get both, the resiliency (if some instances die, you can still handle requests) and the performance (you distribute the load across many instances, and if you have more load, you can add more instances).
Here you also have some optionality in terms of whether those service instances are all interchangeable (each instance loads all rules and has all the facts in the  session) and then you can just round-robin requests to them via a load balancer, or if instances are specialized by the type of request they handle (different instances have different subsets of rules and/or different sets of facts loaded, representing different partitions of your workload) and thus you need a smart router to send different requests to their corresponding processors.

So, to summarize, the session itself cannot be scaled out. But you can duplicate or partition it in the same service or across a server farm. But it all depends on your concrete scenario.
